File: dp/c/circuit_breaker.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

// Service Call Function Pointer Type
typedef bool (*ServiceCall)(void);

typedef enum {
    STATE_CLOSED,     // Variable normal operation (all good)
    STATE_OPEN,       // Circuit open (failing fast)
    STATE_HALF_OPEN   // Testing if service recovers
} CircuitState;

typedef struct {
    CircuitState state;
    int failure_count;
    int failure_threshold;
    time_t last_failure_time;
    int recovery_timeout; // Seconds
} CircuitBreaker;

// --- Circuit Breaker Implementation ---

CircuitBreaker* create_circuit_breaker(int threshold, int timeout_sec) {
    CircuitBreaker* cb = malloc(sizeof(CircuitBreaker));
    cb->state = STATE_CLOSED;
    cb->failure_count = 0;
    cb->failure_threshold = threshold;
    cb->last_failure_time = 0;
    cb->recovery_timeout = timeout_sec;
    return cb;
}

void execute_request(CircuitBreaker* cb, ServiceCall service_func) {
    time_t now = time(NULL);
    
    // Check State
    if (cb->state == STATE_OPEN) {
        if (difftime(now, cb->last_failure_time) > cb->recovery_timeout) {
            printf("[Circuit] Timeout passed, switching to HALF-OPEN.\n");
            cb->state = STATE_HALF_OPEN;
        } else {
            printf("[Circuit] Fast Fail (OPEN). Request blocked.\n");
            return;
        }
    }
    
    // Attempt Call
    bool success = service_func();
    
    if (success) {
        if (cb->state == STATE_HALF_OPEN) {
            printf("[Circuit] Success in HALF-OPEN. Closing circuit.\n");
            cb->state = STATE_CLOSED;
            cb->failure_count = 0;
        } else {
            printf("Service call success.\n");
            // In CLOSED state, we might want to reset failure count on success, 
            // or keep a rolling window. Simple version resets here.
            cb->failure_count = 0; 
        }
    } else {
        printf("Service call FAILED.\n");
        cb->failure_count++;
        cb->last_failure_time = now;
        
        if (cb->state == STATE_HALF_OPEN) {
            printf("[Circuit] Failure in HALF-OPEN. Re-opening circuit.\n");
            cb->state = STATE_OPEN;
        } else if (cb->failure_count >= cb->failure_threshold) {
             printf("[Circuit] Failure limit reached. Opening circuit!\n");
             cb->state = STATE_OPEN;
        }
    }
}

// --- Simulated Service ---

int service_health_score = 100;

bool unreliable_service() {
    // Returns true if "healthy", false if "failed"
    // For demo: fails if health < 50
    return service_health_score >= 50;
}

// --- Usage Example ---

int main() {
    printf("=== Circuit Breaker Pattern (C) ===\n");
    
    // Threshold 3 failures, 2 seconds timeout for demo speed
    CircuitBreaker* cb = create_circuit_breaker(3, 2); 
    
    printf("\n1. Service is healthy (Score 100)\n");
    execute_request(cb, unreliable_service); // Success
    execute_request(cb, unreliable_service); // Success
    
    printf("\n2. Service degrades (Score 40)\n");
    service_health_score = 40;
    
    execute_request(cb, unreliable_service); // Fail 1
    execute_request(cb, unreliable_service); // Fail 2
    execute_request(cb, unreliable_service); // Fail 3 -> Trips breaker
    
    printf("\n3. Requests during OPEN state\n");
    execute_request(cb, unreliable_service); // Fast failed
    execute_request(cb, unreliable_service); // Fast failed
    
    printf("\nWaiting for recovery timeout (3s)...\n");
    #ifdef _WIN32
        _sleep(3000); // ms
    #else
        sleep(3);
    #endif
    
    printf("\n4. Service recovers (Score 90)\n");
    service_health_score = 90;
    
    // First request after timeout should be HALF-OPEN
    execute_request(cb, unreliable_service); // Success -> Resets to CLOSED
    
    execute_request(cb, unreliable_service); // Normal success
    
    free(cb);
    return 0;
}

File: dp/rust/src/circuit_breaker.rs
use std::time::{Duration, Instant};
use std::error::Error;
use std::sync::{Arc, Mutex};
use std::thread;

#[derive(Debug, PartialEq, Clone, Copy)]
enum State {
    Closed,
    Open,
    HalfOpen,
}

struct CircuitBreaker {
    state: State,
    failure_count: u32,
    failure_threshold: u32,
    last_failure: Option<Instant>,
    recovery_timeout: Duration,
}

impl CircuitBreaker {
    fn new(threshold: u32, timeout: Duration) -> Self {
        CircuitBreaker {
            state: State::Closed,
            failure_count: 0,
            failure_threshold: threshold,
            last_failure: None,
            recovery_timeout: timeout,
        }
    }

    fn call<F, T, E>(&mut self, f: F) -> Result<T, String>
    where
        F: FnOnce() -> Result<T, E>,
        E: std::fmt::Display,
    {
        // Check State logic
        if self.state == State::Open {
            if let Some(last) = self.last_failure {
                if last.elapsed() > self.recovery_timeout {
                    println!("[Circuit] Timeout passed. Switching to HALF-OPEN.");
                    self.state = State::HalfOpen;
                } else {
                    return Err("Circuit is OPEN. Fast fail.".to_string());
                }
            }
        }

        // Execute logic
        match f() {
            Ok(val) => {
                if self.state == State::HalfOpen {
                    println!("[Circuit] Success in HALF-OPEN. Resetting to CLOSED.");
                    self.reset();
                } else if self.failure_count > 0 {
                    self.failure_count = 0;
                }
                Ok(val)
            }
            Err(e) => {
                self.record_failure();
                Err(format!("Service execution failed: {}", e))
            }
        }
    }

    fn record_failure(&mut self) {
        self.failure_count += 1;
        self.last_failure = Some(Instant::now());
        println!("[Circuit] Failure recorded. Count: {}", self.failure_count);

        if self.state == State::HalfOpen {
            self.state = State::Open;
            println!("[Circuit] Failed in HALF-OPEN. Re-opening.");
        } else if self.failure_count >= self.failure_threshold {
            self.state = State::Open;
            println!("[Circuit] Threshold reached. Opening circuit.");
        }
    }

    fn reset(&mut self) {
        self.state = State::Closed;
        self.failure_count = 0;
        self.last_failure = None;
    }
}

// Demo function
fn unreliable_service(should_fail: bool) -> Result<String, &'static str> {
    if should_fail {
        Err("Connection Refused")
    } else {
        Ok("Data Payload".to_string())
    }
}

fn main() {
    println!("=== Circuit Breaker Pattern (Rust) ===");

    let cb = Arc::new(Mutex::new(CircuitBreaker::new(3, Duration::from_secs(2))));

    // Helper to run requests
    let run_request = |fail: bool| {
        let mut breaker = cb.lock().unwrap();
        match breaker.call(|| unreliable_service(fail)) {
            Ok(data) => println!("Success: {}", data),
            Err(e) => println!("Error: {}", e),
        }
    };

    println!("\n1. Normal Calls");
    run_request(false);
    
    println!("\n2. Failures");
    run_request(true);
    run_request(true);
    run_request(true); // Trips here

    println!("\n3. Open State (Fast Fail)");
    run_request(false); // Even if true (healthy), it blocks

    println!("\n... Waiting for timeout (2s) ...");
    thread::sleep(Duration::from_secs(2));

    println!("\n4. Recovery (Half-Open)");
    run_request(false); // Should succeed and close
    
    println!("\n5. Back to Normal");
    run_request(false);
}
